indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskClientAsyncImpl.java

```diff
@@ -0,0 +1,648 @@
+    public SeekableStreamRequestBuilder<IntermediateType, FinalType, T> onNotAvailable(final Function<ServiceNotAvailableException, Either<Throwable, T>> fn)
+    {
+      return onException(e -> {
+        if (e instanceof ServiceNotAvailableException) {
+          return fn.apply((ServiceNotAvailableException) e);
+        } else {
+          return Either.error(e);
+        }
+      });
+    }
+
+    /**
+     * Error mapping for {@link ServiceClosedException}, which occurs when a task is not running.
+     */
+    public SeekableStreamRequestBuilder<IntermediateType, FinalType, T> onClosed(final Function<ServiceClosedException, Either<Throwable, T>> fn)
+    {
+      return onException(e -> {
+        if (e instanceof ServiceClosedException) {
+          return fn.apply((ServiceClosedException) e);
+        } else {
+          return Either.error(e);
+        }
+      });
+    }
+
+    /**
+     * Issue the request.
+     */
+    public ListenableFuture<T> go()
+    {
+      final ServiceClient client = makeClient(taskId, retry);
+      final SettableFuture<T> retVal = SettableFuture.create();
+
+      Futures.addCallback(
+          FutureUtils.transform(
+              client.asyncRequest(requestBuilder.timeout(httpTimeout), responseHandler),
+              responseTransformer
+          ),
+          new FutureCallback<T>()
+          {
+            @Override
+            public void onSuccess(@Nullable T result)
+            {
+              retVal.set(result);
+            }
+
+            @Override
+            public void onFailure(Throwable t)
+            {
+              Either<Throwable, T> either = Either.error(t);
+
+              for (final Function<Throwable, Either<Throwable, T>> exceptionMapper : exceptionMappers) {
+                if (!either.isError()) {
+                  break;
+                }
+
+                try {
+                  final Either<Throwable, T> nextEither = exceptionMapper.apply(either.error());
+                  if (nextEither != null) {
+                    either = nextEither;
+                  }
+                }
+                catch (Throwable e) {
+                  // Not expected: on-error function should never throw exceptions. Continue mapping.
+                  log.warn(e, "Failed to map exception encountered while contacting task [%s]", taskId);
+                }
+              }
+
+              if (either.isError()) {
+                retVal.setException(either.error());
+              } else {
+                retVal.set(either.valueOrThrow());
+              }
+            }
+          }
+      );
+
+      return retVal;
+    }
+
+    private ServiceClient makeClient(final String taskId, final boolean retry)
+    {
+      final ServiceRetryPolicy retryPolicy = makeRetryPolicy(taskId, retry);
+      final SeekableStreamTaskLocator locator = new SeekableStreamTaskLocator(taskInfoProvider, taskId);
+      return serviceClientFactory.makeClient(taskId, locator, retryPolicy);
```

Comment:
Would this create a new client on every call? Would it make sense to cache this to help with the preferred location in case of redirects?

Reply:
I did it this way because I didn't want to deal with figuring out how long to keep the clients for, and they're pretty cheap to create. The only state they have is the cache of redirects. But, tasks don't do redirects — only leadery things like Coordinators and Overlords do that. So I think it's OK. I added some comments:

```
      // We're creating a new locator for each request and not closing it. This is OK, since SeekableStreamTaskLocator
      // is stateless, cheap to create, and its close() method does nothing.
      final SeekableStreamTaskLocator locator = new SeekableStreamTaskLocator(taskInfoProvider, taskId);

      // We're creating a new client for each request. This is OK, clients are cheap to create and do not contain
      // state that is important for us to retain across requests. (The main state they retain is preferred location
      // from prior redirects; but tasks don't do redirects.)
      return serviceClientFactory.makeClient(taskId, locator, retryPolicy);

```

Reply:
Thanks!